### PR TITLE
[WIP] Shared string APIs

### PR DESCRIPTION
**Pull request by allevato**
Floating an early sketch of some shared `String` APIs to improve interop code.

Still some open questions, like:
* Should this take advantage of `ContiguousCollection` instead of taking pointers directly (probably)
* Do we need a method to force a shared string into a native one, thus copying it (probably)

cc @milseman

### CODE REVIEW

File: stdlib/public/core/StringStorage.swift [hunk 1]
```diff
@@ -655,6 +682,7 @@ final internal class __SharedStringStorage
   : __SwiftNativeNSString, _AbstractStringStorage {
   internal var _owner: AnyObject?
   internal var start: UnsafePointer<UInt8>
+  internal let deallocator: String.Deallocator
```
Comment (milseman) [inline stdlib/public/core/StringStorage.swift]: Why not have `_owner` be a class whose deinit invokes the closure?

Comment (allevato) [reply stdlib/public/core/StringStorage.swift]: I just didn't know what the `_owner` property was used for 🙂   I've taken the whole `Deallocator` out and made it just a closure in the user-facing API.

File: stdlib/public/core/SharedString.swift [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  ///   also used as the return value for the `withSharedString(_:)` method. The
+  ///   argument is valid only for the duration of the closure's execution.
+  /// - Returns: The return value, if any, of the `body` closure parameter.
+  public func withSharedString<Result>(
+    _ body: (String) throws -> Result
+  ) rethrows -> Result {
+    guard _wholeGuts.isFastUTF8 else {
+      // TODO: What's the right way to get the pointer for a non-fast-UTF-8
+      // string here?
+      return try body(String(self))
```
Comment (milseman) [inline stdlib/public/core/SharedString.swift]: This looks fine to me. I guess the other option would be to form a foreign string, such as a sliced NSString subclass.

File: stdlib/public/core/SharedString.swift [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    return try _wholeGuts.withFastUTF8 { utf8 in
```
Comment (milseman) [inline stdlib/public/core/SharedString.swift]: This isn't small-string correct, but small substrings will form small strings, so eager copy is even faster than using `__SharedStringStorag`. What we'd want to do is check for a contiguous string with a stable pointer.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by karwa** [comment]
ping @milseman @allevato 

Are there any plans to move this along with a pitch? Would you mind if I wrote one?

**Comment by allevato** [comment]
@karwa I don't know what @milseman 's plans are, but I haven't had free time lately to work on this so if you'd like to run with it, feel free!

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**